idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,249 @@
+        Data that is used to build an instance of the Pyomo model and 
+        build the objective function
+    theta_names: list of strings
+        List of Var names to estimate
+    obj_function: function, optional
+        Function used to formulate parameter estimation objective, 
+        generally sum of squared error between measurements and model variables.    
+    tee: bool, optional
+        Indicates that ef solver output should be teed
+    diagnostic_mode: bool, optional
+        If True, print diagnostics from the solver
+    solver_options: dict, optional
+        Provides options to the solver (also the name of an attribute)
+    
+    Returns
+    -------
+    obj: float
+        objective function value for the given obj_function 
+    theta: dictionary
+        Estimated parameters
+    cov: numpy.ndarray
+        Covariance of theta
+    propagation_f: dict
+        error propagation in the objective function with the dictionary key, 'objective' 
+        if the objective function doesn't include any uncertain parameters, return an empty dictionary        
+    propagation_c: dict
+        error propagation in the constraints with the dictionary keys, 'constraints l'
+        where l is the line number. 
+        if no constraint includes uncertain parameters, return an empty dictionary      
+        
+    Raises
+    ------
+    Exception
+        When tee entry is not Boolean
+    Exception
+        When diagnostic_mode entry is not Boolean
+    Exception
+        When solver_options entry is not None and a Dictionary
+    """
+    if not isinstance(tee, bool):
+        raise Exception('tee  must be boolean.')
+    if not isinstance(diagnostic_mode, bool):
+        raise Exception('diagnostic_mode  must be boolean.')    
+    if not solver_options==None:
+        if not isinstance(solver_options, dict):
+            raise Exception('solver_options must be dictionary.')
+
+    parmest_class = parmest.Estimator(model_function, data, theta_names, obj_function,
+                 tee, diagnostic_mode, solver_options)
+    obj, theta, cov = parmest_class.theta_est(calc_cov=True)
+
+    propagation_f, propagation_c =  propagate_uncertainty(model_uncertain, theta, cov, theta_names)
+    return obj, theta, cov, propagation_f, propagation_c
+
+def propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee=False, solver_options=None):
+    """This function calculates gradient vector, expectation, and variance of the objective function 
+    of the model for given estimated optimal parameters and covariance matrix of parameters. It calculates 
+    error propagation of the objective function and constraints by using gradient vector and covariance matrix. 
+    
+    Parameters
+    ----------
+    model_uncertain: function
+        Function that generates an instance of the Pyomo model using
+        'theta' and 'theta_names;  as the input argument
+    theta: dict
```

Comment:
Are the keys of theta the same as theta_names?

Reply:
Yes, 

There is some restriction on theta_names with covariance calculation. 

For example,  `variable_name = ["fs.properties.tau[benzene,toluene]", "fs.properties.tau[toluene,benzene]"]` is ok for both parameter estimation and covariance calculation with `parmest.py `

However, `variable_name = ["fs.properties.tau[benzene, toluene]", "fs.properties.tau[toluene, benzene]"]` is ok for only parameter estimation, not covariance calculation.

Reply:
@JanghoPark-LBL Let's add an assert statement to check for a space in the parameter name and warn the user

Reply:
I added this warning.

Reply:
Update your to be created `clean_variable_name` function to also remove spaces.

Reply:
Suggestion from @dangunter: Update to `clean_variable_name` to log warnining message when it makes changes.

Reply:
Also write a few unit tests for the clean_variable_name function.

Reply:
Did this get resolved?

Reply:
Yes, I added unit tests for the clean_variable_name function.